### PR TITLE
Use oak_crypto in Oak Functions

### PR DESCRIPTION
**Pull request by ipetr0v**
This PR makes Oak Functions and Oak Client use `oak_crypto` for message communication with the client.

Currently actual hybrid encryption is commented out (with a TODO(#3642) added), because we also need to implement Java hybrid encryption to not break tests.

Ref https://github.com/project-oak/oak/issues/3767 https://github.com/project-oak/oak/issues/3640
Fixes https://github.com/project-oak/oak/issues/3643

This PR is a split from https://github.com/project-oak/oak/pull/3790

### CODE REVIEW

**Comment by tiziano88** [comment]
I think it's fine to mark those tests as ignored and add a TODO with the corresponding bug. But please make sure you have a good idea of when they will be fixed again, we don't want to leave them ignored for more than a few days I guess.

**Comment by ipetr0v** [comment]
For now commented out encryption with a TODO(https://github.com/project-oak/oak/issues/3642)

File: java/src/main/java/com/google/oak/client/OakGrpcClient.java [hunk 1]
```diff
@@ -76,17 +80,30 @@ public void onCompleted() {
     };
     StreamObserver<RequestWrapper> requestObserver = stream.apply(responseObserver);
 
+    HpkeRequest hpkeRequest =
```
Comment (tiziano88) [inline java/src/main/java/com/google/oak/client/OakGrpcClient.java:83]: I believe there should be a TODO around here, right?

File: oak_client/src/lib.rs [hunk 2]
```diff
@@ -22,34 +22,92 @@ pub mod proto {
...
 
+const EMPTY_ASSOCIATED_DATA: &[u8] = b"";
+
 /// Client for connecting to Oak.
 /// Represents a Relying Party from the RATS Architecture:
 /// <https://www.rfc-editor.org/rfc/rfc9334.html#name-relying-party>
 pub struct OakClient<T>
 where
-    T: micro_rpc::AsyncTransport<Error = anyhow::Error>,
+    T: micro_rpc::AsyncTransport<Error = anyhow::Error> + AsyncEvidenceProvider,
```
Comment (tiziano88) [inline oak_client/src/lib.rs:41]: Why is this needed here? AFAICT it's only needed to create the client, and never after that, so it can just be an additional bound for the create method?

File: oak_client/src/lib.rs [hunk 3]
```diff
@@ -22,34 +22,92 @@ pub mod proto {
...
 {
-    pub async fn create(transport: T) -> anyhow::Result<Self> {
+    pub async fn create(mut transport: T) -> anyhow::Result<Self> {
         // TODO(#3442): Fetch enclave public key.
         // TODO(#3641): Implement client-side attestation verification.
-        Ok(Self { transport })
+        let evidence = transport
+            .get_evidence()
```
Comment (tiziano88) [inline oak_client/src/lib.rs:55]: It seems overkill to pass an entire AsyncEvidenceProvider -- why not just pass the evidence, obtained by the caller? Alongside the transport (which would not require the extra trait bound at all).

File: oak_client/src/lib.rs [hunk 4]
```diff
@@ -22,34 +22,92 @@ pub mod proto {
...
-            .context("couldn't send request")
+            .map_err(|error| anyhow!("couldn't send request: {:?}", error))?;
+
+        // Deserialize and decrypt response.
+        let response = HpkeResponse::decode(serialized_response.as_ref())
+            .map_err(|error| anyhow!("couldn't deserialize response: {:?}", error))?;
+        let encrypted_message = response
+            .encrypted_message
+            .context("response doesn't contain encrypted message")?;
+        // The resulting encryptor for consequent requests is discarded because we don't send
```
Comment (tiziano88) [inline oak_client/src/lib.rs:102]:
```suggestion
        // The resulting encryptor for subsequent requests is discarded because we don't send
```

File: oak_client/src/lib.rs [hunk 5]
```diff
@@ -22,34 +22,92 @@ pub mod proto {
...
+        // another message to the stream.
```
Comment (tiziano88) [inline oak_client/src/lib.rs:103]: Why do we need to re-create the encryptor every time? Would it not be possible to have an API that allows reusing it? Otherwise we end up with so many encryptors if we were to actually do a stream. Not for this PR, but perhaps we can follow up separately?

File: oak_client/src/lib.rs [hunk 6]
```diff
@@ -22,34 +22,92 @@ pub mod proto {
...
-        self.transport
-            .invoke(request_body)
+        // Create encryptor.
+        let (serialized_encapsulated_public_key, request_encryptor) = self
+            .crypto_provider
+            .create_encryptor()
+            .context("couldn't create encryptor")?;
+
+        // Encrypt and serialize request.
+        let (request_ciphertext, response_decryptor) = request_encryptor
```
Comment (tiziano88) [inline oak_client/src/lib.rs:73]: Why is the decryptor returned by the encryptor? Would it not be possible to return both encryptor and decryptor at the same time from the crypto provider?

File: oak_crypto/src/aead.rs [hunk 7]
```diff
@@ -34,7 +34,7 @@ pub(crate) type AeadNonce = [u8; AEAD_NONCE_SIZE_BYTES];
 
 /// Encrypts `plaintext` with associated data using AES-GCM encryption scheme.
 /// Note: the corresponding associated data is NOT encrypted.
-pub(crate) fn encrypt(
+pub(crate) fn _encrypt(
```
Comment (tiziano88) [inline oak_crypto/src/aead.rs:37]: Why was this (and others) renamed?

**Review by tiziano88 — COMMENTED** [review]
Awesome! A few comments so far

File: oak_crypto/src/aead.rs (hunk 7 above)
Comment (ipetr0v) [reply oak_crypto/src/aead.rs:37]: Because they were producing `unused function` errors

File: oak_client/src/lib.rs (hunk 5 above)
Comment (ipetr0v) [reply oak_client/src/lib.rs:103]: I'm currently refactoring crypto API in a different PR

File: oak_client/src/lib.rs (hunk 3 above)
Comment (ipetr0v) [reply oak_client/src/lib.rs:55]: This would mean that we'll need to create a second `EvidenceRequestor` client which only requests the evidence instead of hiding all of the logic inside `OakClient`. And this `EvidenceRequestor` client would also consume the gRPC transport.

Comment (tiziano88) [reply oak_client/src/lib.rs:55]: Is it not possible to invoke it without consuming it? I thought a mut reference would be enough.

Comment (ipetr0v) [reply oak_client/src/lib.rs:55]: This would still mean that we won't have encapsulation of the client logic in a single object.

File: oak_client/src/lib.rs (hunk 4 above)
Comment (ipetr0v) [reply oak_client/src/lib.rs:102]: Done

File: oak_functions_launcher/tests/integration_test.rs [hunk 8]
```diff
@@ -98,15 +105,61 @@ async fn test_launcher_looks_up_key() {
...
+    let invoke_response = client
         .invoke(&invoke_request)
         .await
-        .expect("Failed to receive response.");
-
-    assert!(response.is_ok());
-    assert_eq!(b"test_value".to_vec(), response.unwrap().body);
+        .expect("couldn't to receive response");
+    assert!(invoke_response.is_ok());
+    // let response = client
```
[the remark this answers is not in the document]
Comment (ipetr0v) [reply oak_functions_launcher/tests/integration_test.rs]: Done
